Code Quality: Global Constants, Organized Imports, Simplified Set Creation

realshouzy opened the pull request:
Made some (minor) adjustments:
1. Replaced the functions ``pip_cmd`` and ``version_epilog`` with global constants ``PIP_CMD`` and ``VERSION_EPILOG``, since function calls in Python are quite expensive (although it probably won’t make much of a difference) and I find it to be more readable. All global constants are now also declared  at the top of the module.
2. Removed unnecessary ``PY3`` constant assignment
3. Sorted imports according to [PEP8](https://peps.python.org/pep-0008/#imports).
4. Simplified set creation for ``LIST_ONLY`` and ``INSTALL_ONLY``
5. Renamed ``format`` to ``format_`` to avoid clashing with built-in function
6. Replaced existing ``.gitignore`` with a template for Python projects

realshouzy committed:
Make pip_cmd and verison_epilog constants

Sort imports according to pep8

Remove unnecessary PY3 assignment

realshouzy committed:
Simplify set creation of LIST_ONLY and INSTALL_ONLY

realshouzy committed:
Rename 'format' to 'format_' to avoid clash with Python built-in function

realshouzy committed:
Uncapitalize code comment

realshouzy committed:
Use literal syntax for set creation

jgonggrijp commented on .gitignore:
.gitignore templates are useful for getting up to speed with a new project. For an established project that already has a custom .gitignore, they just add weight.

jgonggrijp commented on pip_review/__main__.py:
```diff
@@ -49,24 +50,76 @@ def check_output(*args, **kwargs):
 '''
 
 # parameters that pip list supports but not pip install
-LIST_ONLY = set('l local path format not-required exclude-editable include-editable'.split())
+LIST_ONLY = {
+    'l',
+    'local',
+    'path',
+    'format',
+    'not-required',
+    'exclude-editable',
+    'include-editable',
+}
```
You wrote exactly the same set with a different notation. I suppose the notation you chose is more canonical or perhaps "cleaner". But does that justify introducing another change in the git history, which future developers might need to search with a `git blame` or a `git bisect` in order to trace why the contents of this set are the way they are?

jgonggrijp commented on pip_review/__main__.py:
```diff
@@ -49,24 +50,76 @@ def check_output(*args, **kwargs):
 
 # parameters that pip install supports but not pip list
-INSTALL_ONLY = set('c constraint no-deps t target platform python-version implementation abi root prefix b build src U upgrade upgrade-strategy force-reinstall I ignore-installed ignore-requires-python no-build-isolation use-pep517 install-option global-option compile no-compile no-warn-script-location no-warn-conflicts no-binary only-binary prefer-binary no-clean require-hashes progress-bar'.spl ...
+INSTALL_ONLY = {
+    'c',
+    'constraint',
+    'no-deps',
+    't',
+    'target',
+    'platform',
+    'python-version',
+    'implementation',
+    'abi',
+    'root',
+    'prefix',
+    'b',
+    'build',
+    'src',
+    'U',
+    'upgrade',
+    'upgrade-strategy',
+    'force-reinstall',
+    'I',
+    'ignore-installed',
+    'ignore-requires-python',
+    'no-build-isolation',
+    'use-pep517',
+    'install-option',
+    'global-option',
+    'compile',
+    'no-compile',
+    'no-warn-script-location',
+    'no-warn-conflicts',
+    'no-binary',
+    'only-binary',
+    'prefer-binary',
+    'no-clean',
+    'require-hashes',
+    'progress-bar',
+}
```
Also, does a constant really need to occupy 37 lines just because it is cleaner?

jgonggrijp reviewed (changes requested):
I think some of these changes would make sense if included in another pull request, but nothing in this branch seems to be solving a real problem. On the whole, it looks like a bunch of changes for the sake of making changes.

Convince me: why do these changes (or some of them at least) *need* to be made?

realshouzy commented:
Thanks for the feedback, I understand that some of these changes should be included in separate pull requests (please excuse my ignorance, I’m still relatively new to this open source stuff).

So firstly, I realize that the ``.gitignore`` template is unnecessary weight, although I would at least include ``venv/`` to save future developers the hassle of having to add it to the ``.gitignore`` themselves.

Regarding the other proposed changes, they aim to enhance the quality of code adhering to PEP8 and the [Python Code Quality Authority](https://meta.pycqa.org/introduction.html). Why do these *need* to be made? Well, while opinions may differ, I believe they improve the willingness of potential contributors to collaborate and support ongoing project maintenance. While they might not immediately simplify the git history, they greatly improve code readability and comprehensibility. In conjunction with thorough code comments and documentation, developers can grasp the code's purpose without delving into the git history. Moreover, cleaner code tends to minimize potential bugs.
For instance, simplifying the set creation reduces code complexity. Initially, the set creation using a string as the iterable might confuse readers (at least it did confuse me) until they notice the subsequent ``.split()`` method call. This can easily be missed, especially when having to scrolling horizontally. Moreover, one would also need to know, what ``.split()`` actual does. Utilizing literal syntax and breaking the statement across multiple lines allows developers to quickly understand the entire statement's functionality at a glance, whether skimming it or examining it closely.

I hope this explanation makes sense. Perhaps a compromise is the way to go. I really value this tool and believe that enhancing its code quality will mitigate project neglect.

jgonggrijp commented:
Alright, well argumented. If you leave out [2c03879](https://github.com/jgonggrijp/pip-review/pull/109/commits/2c03879266abc291ceac5459b97be24bb282bea8) (the `.gitignore` commit) and wrap those set literals to put as many strings per line as can fit within 79 columns, I'll accept it. It's OK to add `venv/` to the `.gitignore`, if you use that name for your venvs or know someone who does.

It would be great if this helps to attract new maintainers!

realshouzy commented:
When it comes to the maximum line length, there are various conventions to consider. PEP8 suggests 79 characters, while other standards like the [Python Black formatter](https://github.com/psf/black) opt for 88, but there is also 99, 100, or 120 characters. Many (including myself) find 88 characters to be a balanced choice – neither too lengthy nor too restrictive. Ultimately, the decision rests with you, as it's your project. So it’s gonna be 79?

jgonggrijp commented:
Short answer: yes, 79 please.

Longer answer with rationale:

To be honest, I find the 79 character rule a bit unnatural and always configure my editors to wrap lines at 80 characters instead. That said, I do think that restricting line length is a good idea, because not everyone likes to use their editor at full screen width and even when they do, I doubt that everyone finds it comfortable to turn their heads left and right while reading long lines. It's also just easier to find the start of the next line if the text column is not too wide.

I consider 79 or 80 a compromise already. 70 is most ergonomic for natural language text; we add 10 columns when programming to account for indentation. If you use more than 2.5 levels of indentation on average, you should probably deforest your code or use more functional style.

If we look at the `INSTALL_ONLY` set in particular, it started out as a single line of 404 characters. That is obviously too long. You initially changed it into 37 lines of about 17 characters on average, with just one string per line. That (I think) is also obviously too long, but in the vertical direction. If my counting is right, keeping the set in literal notation but wrapping it to 79 columns will reduce this to 9 lines, which is a huge improvement. Wrapping it to 88 columns will not save any more lines, but force some people to scroll horizontally or have their editor soft-wrap the lines longer than 80 characters, with the end result of taking up more vertical space again.

realshouzy commented:
Agreed; should I create a new branch with its respective pull request?

jgonggrijp commented:
No, you can rebase this branch and then force-push. You can go back to the old version of your branch with `git reflog` in case you mess up.

realshouzy commented:
Alright, thanks for the tip, I'll do that as soon as I find the time (should be within the next two days or so).

realshouzy committed:
Add venv/ to .gitignore

realshouzy committed:
Format set creations within 79 columns

realshouzy commented:
It should be done now.

jgonggrijp commented:
I ran the tests, they still pass. Congratulations with your first contribution to this project, and thanks!

realshouzy commented:
You're welcome and thank you!
